File: QC.R
library(Seurat)
library(sceasy)
library(reticulate)
library(tidyverse)
library(argparse)
source("functions_ashr.R")
library(conflicted)
conflict_prefer("select", "dplyr")
conflict_prefer("filter", "dplyr")
conflict_prefer("rename", "dplyr")
conflicts_prefer(base::setdiff)

use_condaenv('env_nf')

parser <- ArgumentParser()
parser$add_argument("--cells", type = "character", help = "Cell type/line", required = TRUE)
args <- parser$parse_args()
cells <- args$cells

read_perturb_h5ad <- function(cells) {
  if(cells == "Jurkat") {
    h5ad_file_name <- "jurkat_raw_singlecell_01.h5ad"
  } else if(cells == "K562_GenomeWide") {
    h5ad_file_name <- "K562_gwps_raw_singlecell.h5ad"
  } else if(cells == "HepG2") {
    h5ad_file_name <- "hepg2_raw_singlecell_01.h5ad"
  } else if(cells == "RPE1") {
    h5ad_file_name <- "rpe1_raw_singlecell_01.h5ad"
  }
  return(
    sceasy::convertFormat(file.path("/rds/project/rds-csoP2nj6Y6Y/biv22/perturb_pseudobulk_de/data", h5ad_file_name), from="anndata", to="seurat")
  )
}

seurat_obj <- read_perturb_h5ad(cells)

DEGs <- list.files(file.path("/rds/project/rds-csoP2nj6Y6Y/biv22/data/perturb", cells), full.names = T)

pairs <- extract_DEGs(DEGs, cores = 32, bulk=T)$perturbation_effect_df %>%
  select(perturbation, effect_ensg)

gene_QC_metrics <- function(seurat_object, pairs, expression_threshold = 0, min_cells = 7) {
  expr_mat <- GetAssayData(seurat_object, slot = "data")
  meta <- seurat_object@meta.data
  
  # Filter pairs to valid ones
  valid_genes <- rownames(expr_mat)
  valid_perts <- unique(meta$gene)
  pairs <- pairs %>%
    filter(effect_ensg %in% valid_genes, perturbation %in% valid_perts)
  
  cat("Total valid pairs:", nrow(pairs), "\n")
  
  # Get control expression
  ctrl_cells <- rownames(meta)[meta$gene == "non-targeting"]
  ctrl_expr_mat <- expr_mat[, ctrl_cells, drop = FALSE]
  ctrl_counts <- Matrix::rowSums(ctrl_expr_mat > expression_threshold)
  ctrl_df <- tibble(gene = names(ctrl_counts), n_expr_ctrl = ctrl_counts)
  
  # Group treatment cells by perturbation
  pert_cell_index <- split(rownames(meta), meta$gene)
  pert_cell_index <- pert_cell_index[names(pert_cell_index) != "non-targeting"]
  
  trt_names <- names(pert_cell_index)
  total_perts <- length(trt_names)
  
  # Process each perturbation
  trt_counts_list <- mclapply(seq_along(trt_names), function(i) {
    pert <- trt_names[i]
    cat(sprintf("Processing %d / %d: %s\n", i, total_perts, pert))
    
    pert_cells <- pert_cell_index[[pert]]
    trt_expr_mat <- expr_mat[, pert_cells, drop = FALSE]
    counts <- Matrix::rowSums(trt_expr_mat > expression_threshold)
    
    tibble(gene = names(counts), perturbation = pert, n_expr_trt = counts)
  }, mc.cores = 32, mc.preschedule = FALSE)
  
  trt_counts_df <- bind_rows(trt_counts_list)
  
  # Join with pair list and control counts
  df <- pairs %>%
    rename(gene = effect_ensg) %>%
    left_join(trt_counts_df, by = c("perturbation", "gene")) %>%
    left_join(ctrl_df, by = "gene") %>%
    filter(!is.na(n_expr_trt), !is.na(n_expr_ctrl)) %>%
    mutate(QC_pass = ifelse(n_expr_trt >= min_cells & n_expr_ctrl >= min_cells, "Pass", "Fail"))
  
  return(df)
}

gene_QC <- gene_QC_metrics(seurat_obj, pairs)


DRGs <- read_csv(file.path("/rds/project/rds-csoP2nj6Y6Y/biv22/data/perturb/pairs", cells, "perturbation_pairs_eff.csv"))

map_ensg_to_symbol <- function(gene_QC) {
  gene.ensg <- gene_QC$gene %>% unique
  
  gene.symbols <- mapIds(org.Hs.eg.db, keys = gene.ensg,
                            column = "SYMBOL", keytype = "ENSEMBL",
                            multiVals = "first")
  
  gene.symbols <- gene.symbols %>% na.omit
  
  gene_QC <- gene_QC %>%
    mutate(effect = gene.symbols[gene])
  
  return(gene_QC)
}

gene_QC <- gene_QC %>% map_ensg_to_symbol %>% filter(perturbation != effect)


gene_QC <- left_join(
  gene_QC,
  DRGs,
  by = c("perturbation", "effect")
) %>%
  select(perturbation, effect, n_expr_trt, n_expr_ctrl, avg_log2FC, p_val_adj)

write_csv(gene_QC, file.path("/rds/project/rds-csoP2nj6Y6Y/biv22/data/perturb_QC", cells, "gene_QC.csv"))

File: QC.py
#!/usr/bin/env python
import scanpy as sc
import pandas as pd
import numpy as np
import argparse
from pathlib import Path
from tqdm import tqdm
import mygene


def map_ensg_to_symbol(df, gene_col='effect'):
    """
    Maps ENSEMBL IDs in `gene_col` to gene symbols using mygene.

    Returns a DataFrame with:
        - 'effect_ensg' = original ENSEMBL
        - 'effect' = gene symbol
    """
    mg = mygene.MyGeneInfo()
    ensg_list = df[gene_col].unique().tolist()

    print(f"Mapping {len(ensg_list)} ENSG IDs to symbols...")
    results = mg.querymany(ensg_list, scopes='ensembl.gene', fields='symbol', species='human')

    # Build mapping dictionary
    symbol_map = {r['query']: r.get('symbol', None) for r in results if 'symbol' in r}

    # Rename original column and add mapped symbol
    df = df.copy()
    df = df.rename(columns={gene_col: 'effect_ensg'})
    df['effect'] = df['effect_ensg'].map(symbol_map)

    # Remove rows where symbol mapping failed
    df = df.dropna(subset=['effect'])
    return df

def extract_DEGs(DEG_dir, significant=True, bulk=True):
    DEG_dir = Path(DEG_dir)
    DEG_files = list(DEG_dir.glob("*.tsv"))
    if len(DEG_files) == 0:
        raise FileNotFoundError(f"No DEG TSV files found in {DEG_dir}")
    
    dfs = []
    for f in DEG_files:
        DEGs = pd.read_csv(f, sep="\t", index_col=0)
        perturbed_gene = f.stem
        if bulk:
            # mimic column renaming from R
            DEGs = DEGs.rename(columns={"log2FoldChange": "avg_log2FC",
                                        "padj": "p_val_adj"})
        DEGs = DEGs.reset_index().rename(columns={"gene": "effect"})
        DEGs['perturbation'] = perturbed_gene
        dfs.append(DEGs)
    
    all_DEGs = pd.concat(dfs, ignore_index=True)
    
    if significant:
        all_DEGs = all_DEGs[all_DEGs['p_val_adj'] < 0.05]
    
    # Keep only columns needed for QC
    pairs = all_DEGs[['perturbation', 'effect']]
    return pairs

# -----------------------------
# ARGUMENTS
# -----------------------------
parser = argparse.ArgumentParser()
parser.add_argument("--cells", type=str, required=True, help="Cell type/line")
args = parser.parse_args()
cells = args.cells

# -----------------------------
# FILE PATHS
# -----------------------------
DATA_DIR = Path("/rds/project/rds-csoP2nj6Y6Y/biv22/perturb_pseudobulk_de/data")
PERTURB_DIR = Path(f"/rds/project/rds-csoP2nj6Y6Y/biv22/data/perturb/{cells}")
OUT_DIR = Path(f"/rds/project/rds-csoP2nj6Y6Y/biv22/data/perturb_QC/{cells}")
OUT_DIR.mkdir(exist_ok=True, parents=True)

h5ad_files = {
    "Jurkat": "jurkat_raw_singlecell_01.h5ad",
    "K562_GenomeWide": "K562_gwps_raw_singlecell.h5ad",
    "K562_essential": "K562_essential_raw_singlecell_01.h5ad",
    "HepG2": "hepg2_raw_singlecell_01.h5ad",
    "RPE1": "rpe1_raw_singlecell_01.h5ad"
}

# -----------------------------
# LOAD DATA
# -----------------------------
adata = sc.read_h5ad(DATA_DIR / h5ad_files[cells])

# Ensure perturbation is in obs
if 'gene' not in adata.obs:
    raise ValueError("Expected 'gene' column in adata.obs for perturbations")

# -----------------------------
# LOAD DEGs / extract pairs
# -----------------------------
# Assumes you have an equivalent extract_DEGs output as CSVs in PERTURB_DIR
DEG_dir = f"/rds/project/rds-csoP2nj6Y6Y/biv22/data/perturb/{cells}"
pairs = extract_DEGs(DEG_dir, significant=False, bulk=True)
pairs = map_ensg_to_symbol(pairs)


# -----------------------------
# GENE QC FUNCTION
# -----------------------------
def gene_qc(adata, pairs, expression_threshold=0, min_cells=7):
    # Filter pairs to valid genes and perturbations
    valid_genes = adata.var_names
    valid_perts = adata.obs['gene'].unique()
    pairs = pairs[pairs['effect_ensg'].isin(valid_genes) & pairs['perturbation'].isin(valid_perts)]
    print(f"Total valid pairs: {len(pairs)}")

    # CONTROL CELLS
    ctrl_cells = adata.obs_names[adata.obs['gene'] == 'non-targeting']
    ctrl_counts = np.array((adata[ctrl_cells, :].X > expression_threshold).sum(axis=0)).flatten()
    ctrl_df = pd.DataFrame({'gene': adata.var_names, 'n_expr_ctrl': ctrl_counts})

    # TREATMENT CELLS
    trt_dfs = []
    for pert in tqdm([p for p in valid_perts if p != 'non-targeting']):
        pert_cells = adata.obs_names[adata.obs['gene'] == pert]
        counts = np.array((adata[pert_cells, :].X > expression_threshold).sum(axis=0)).flatten()
        trt_dfs.append(pd.DataFrame({
            'gene': adata.var_names,
            'perturbation': pert,
            'n_expr_trt': counts
        }))
    trt_counts_df = pd.concat(trt_dfs, ignore_index=True)

    # JOIN AND QC FLAG
    df = pairs.rename(columns={'effect_ensg': 'gene'})
    df = df.merge(trt_counts_df, on=['perturbation','gene'], how='left')
    df = df.merge(ctrl_df, on='gene', how='left')
    df = df.dropna(subset=['n_expr_trt','n_expr_ctrl'])
    df['QC_pass'] = np.where((df['n_expr_trt'] >= min_cells) & (df['n_expr_ctrl'] >= min_cells),
                             'Pass', 'Fail')
    return df

gene_QC = gene_qc(adata, pairs)

# -----------------------------
# MAP ENSG -> SYMBOL
# -----------------------------
mg = mygene.MyGeneInfo()
symbols = mg.querymany(gene_QC['gene'].unique(), scopes='ensembl.gene', fields='symbol', species='human')
symbol_map = {s['query']: s.get('symbol', None) for s in symbols}
gene_QC['effect'] = gene_QC['gene'].map(symbol_map)
# Don't filter perturbation == effect. Won't be able to estimate KD efficiency
# gene_QC = gene_QC[gene_QC['perturbation'] != gene_QC['effect']]

# -----------------------------
# SAVE RESULTS
# -----------------------------
gene_QC.to_csv(OUT_DIR / "gene_QC.csv", index=False)
print(f"QC done, saved to {OUT_DIR / 'gene_QC.csv'}")

